### PR TITLE
Emitting metrics for DLQ message count

### PR DESCRIPTION
**Pull request by prathyushpv**
## What changed?
Added a DLQMEtricsEmitter which will emit DLQ message count every 3 hours.

## Why?
We alredy have a counter metric for DLQ writes. This will increment each time a new message is written to DLQ. This can be used to create an alarm that fires when a message is added to DLQ. But we also need to know if DLQ messages are inspected and cleared on time. This is not possible with the existing metric. This new metric DLQMessageCount will help in monitoring the current number of messages in DLQ. This metric will only be emitted from the history service instance which hosts shard numbered 1.

I also considered adding a gauge metric for DLQ message count. If the history service restarts, we will lose this count.

## How did you test it?
Unit test

## Potential risks
None

### CODE REVIEW

File: common/persistence/dlq_metrics_emitter.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/service/history/tasks"
+)
+
+const (
+	emitDLQMetricsInterval = 1 * time.Hour
```
Comment (yycptt) [inline common/persistence/dlq_metrics_emitter.go:40]: I think we need to make this interval based on history service ring size. Or make only one history pod emit the metric (e.g. look up a fixed key or the host has the smallest ip/name etc.).

Comment (prathyushpv) [reply common/persistence/dlq_metrics_emitter.go:39]: Yeah, that makes sense! I have added a condition to emit metrics from the history instance which hosts shard number 0.

File: service/history/shard/controller_impl.go [hunk 2]
```diff
@@ -132,6 +135,7 @@ func (c *ControllerImpl) Start() {
 	}
 
 	c.ownership.start(c)
+	c.dlqMetricsEmitter.Start()
```
Comment (yycptt) [inline service/history/shard/controller_impl.go:138]: hmmm don't think it belong there. Maybe move it to history handler?

File: common/persistence/dlq_metrics_emitter.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+				category, err := GetHistoryTaskQueueCategoryID(q.QueueName)
+				if err != nil {
+					s.logger.Error("Failed to process DLQ queue name", tag.Error(err))
+				}
+				messageCounts[category] += q.MessageCount
+			}
+			for id, count := range messageCounts {
+				category, ok := categories[id]
+				if !ok {
+					s.logger.Error("Failed to find category from ID ", tag.TaskCategoryID(id))
```
Comment (yycptt) [inline common/persistence/dlq_metrics_emitter.go:124]:
```suggestion
					s.logger.Error("Failed to find category from ID", tag.TaskCategoryID(id))
```

File: common/persistence/dlq_metrics_emitter.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+			}
+		}
+	}
+}
+
+func (s *DLQMetricsEmitter) getDLQList() ([]QueueInfo, error) {
+	var queues []QueueInfo
+	var nextPageToken []byte
+	for {
+		resp, err := s.historyTaskQueueManager.ListQueues(context.Background(), &ListQueuesRequest{
```
Comment (yycptt) [inline common/persistence/dlq_metrics_emitter.go:136]: we should mark the request as low priority. `SystemPreemptableCallerInfo`

File: service/history/shard/controller_impl.go (hunk 2 above)
Comment (prathyushpv) [reply service/history/shard/controller_impl.go]: Moved.

File: common/persistence/dlq_metrics_emitter.go (hunk 4 above)
Comment (prathyushpv) [reply common/persistence/dlq_metrics_emitter.go]: Done.

**Review by yycptt — APPROVED** [review]

**Pull request merged**